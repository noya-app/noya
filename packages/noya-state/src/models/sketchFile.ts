import Sketch from '@sketch-hq/sketch-file-format-ts';
import { SketchFile } from 'noya-sketch-file';
import { Models } from '..';
import meta from './meta';
import { createDocument } from './document';

export function createSketchFile(): SketchFile {
  const pages = [Models.page];

  return {
    document: createDocument(),
    images: {},
    meta,
    pages,
    user: {
      document: {
        pageListHeight: 0,
        pageListCollapsed: Sketch.NumericalBool.False,
      },
      ...Object.fromEntries(
        pages.map((page) => [
          page.do_objectID,
          {
            scrollOrigin: '{0, 0}',
            zoomValue: 1,
          },
        ]),
      ),
    },
  };
}
