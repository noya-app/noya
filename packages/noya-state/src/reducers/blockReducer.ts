import { CanvasKit } from 'canvaskit';
import Sketch from 'noya-file-format';
import type { ApplicationState } from './applicationReducer';
import { layerPropertyReducer } from './layerPropertyReducer';
import { symbolsReducer } from './symbolsReducer';

export type BlockContent = {
  blockText: string;
  normalizedText?: string;
  overrides?: Sketch.OverrideValue[];
  symbolId?: string;
};

export type OverriddenBlockContent = {
  layerId: string;
  blockContent: BlockContent;
};

export type BlockAction =
  | [type: 'setBlockContent', layerId: string, content: BlockContent];

export function blockReducer(
  state: ApplicationState,
  action: BlockAction,
  CanvasKit: CanvasKit,
): ApplicationState {
  switch (action[0]) {
    case 'setBlockContent': {
      const [, id, { blockText, normalizedText, overrides, symbolId }] = action;

      if (symbolId) {
        state = symbolsReducer(state, [
          'setSymbolInstanceSource',
          symbolId,
          'preserveCurrent',
        ]);
      }

      state = layerPropertyReducer(
        state,
        ['setBlockText', id, blockText, normalizedText],
        CanvasKit,
      );

      state = (overrides ?? []).reduce((result, override) => {
        return symbolsReducer(result, [
          'setOverrideValue',
          [id],
          override.overrideName,
          override.value,
        ]);
      }, state);

      return state;
    }
    default:
      return state;
  }
}
