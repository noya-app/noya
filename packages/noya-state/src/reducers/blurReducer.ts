import { Sketch } from '@noya-app/noya-file-format';
import { clamp } from '@noya-app/noya-utils';
import produce from 'immer';
import { SketchModel } from 'noya-sketch-model';
import { SetNumberMode } from '../types';

export type BlurAction =
  | [type: 'setBlurEnabled', isEnabled: boolean]
  | [type: 'setBlurType', value: Sketch.BlurType]
  | [type: 'setBlurRadius', value: number, mode: SetNumberMode]
  | [type: 'setBlurSaturation', value: number, mode: SetNumberMode];

export function blurReducer(
  state: Sketch.Blur | undefined,
  action: BlurAction,
): Sketch.Blur {
  state = state ?? SketchModel.blur();

  switch (action[0]) {
    case 'setBlurEnabled': {
      const [, isEnabled] = action;

      return produce(state, (draft) => {
        draft.isEnabled = isEnabled;
      });
    }
    case 'setBlurType': {
      const [, value] = action;

      return produce(state, (draft) => {
        draft.type = value;
      });
    }
    case 'setBlurRadius': {
      const [, amount, mode = 'replace'] = action;

      return produce(state, (draft) => {
        const newValue =
          mode === 'replace' ? amount : (draft.radius ?? 0) + amount;

        draft.radius = clamp(newValue, 0, 50);
      });
    }
    case 'setBlurSaturation': {
      const [, amount, mode = 'replace'] = action;

      return produce(state, (draft) => {
        const newValue =
          mode === 'replace' ? amount : draft.saturation + amount;

        draft.saturation = clamp(newValue, 0, 2);
      });
    }
    default:
      return state;
  }
}
