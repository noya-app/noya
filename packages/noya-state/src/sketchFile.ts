import { Sketch } from '@noya-app/noya-file-format';
import { SketchFile } from 'noya-sketch-file';
import { SketchModel } from 'noya-sketch-model';

export function createSketchFile(
  page: Sketch.Page = SketchModel.page(),
): SketchFile {
  return {
    document: SketchModel.document(),
    images: {},
    meta: SketchModel.meta(),
    pages: [page],
    user: SketchModel.user({
      [page.do_objectID]: {
        scrollOrigin: '{0, 0}',
        zoomValue: 1,
      },
    }),
  };
}
