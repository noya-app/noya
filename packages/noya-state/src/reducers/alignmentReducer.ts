import produce from 'immer';
import { IndexPath } from 'tree-visit';

import Sketch from 'noya-file-format';
import { Bounds, createBounds, transformRect } from 'noya-geometry';
import { sum } from 'noya-utils';
import * as Layers from '../layers';
// import {
//   getCurrentPage,
//   getCurrentPageIndex,
//   getLayerTransformAtIndexPath,
//   getSelectedLayerIndexPaths,
//   getSelectedRect,
// } from '../selectors/selectors';
import { accessPageLayers, ApplicationState } from './applicationReducer';

export type AlignmentAction =
  | [type: 'distributeLayers', placement: 'horizontal' | 'vertical']
  | [
      type: 'alignLayers',
      placement:
        | 'left'
        | 'centerHorizontal'
        | 'right'
        | 'top'
        | 'centerVertical'
        | 'bottom',
    ];

// export function alignmentReducer(
//   state: ApplicationState,
//   action: AlignmentAction,
// ): ApplicationState {
//   switch (action[0]) {
//     case 'distributeLayers': {
//       const page = getCurrentPage(state);
//       const pageIndex = getCurrentPageIndex(state);
//       const layerIndexPaths = getSelectedLayerIndexPaths(state);
//       const selectedRect = getSelectedRect(state);
//       const [, axis] = action;

//       return produce(state, (draft) => {
//         const layers = accessPageLayers(draft, pageIndex, layerIndexPaths);
//         const combinedWidths = sum(layers.map((layer) => layer.frame.width));
//         const combinedHeights = sum(layers.map((layer) => layer.frame.height));
//         const differenceWidth = selectedRect.width - combinedWidths;
//         const differenceHeight = selectedRect.height - combinedHeights;
//         const gapX = differenceWidth / (layers.length - 1);
//         const gapY = differenceHeight / (layers.length - 1);
//         const sortBy = axis === 'horizontal' ? 'midX' : 'midY';

//         // Bounds are all transformed to the page's coordinate system
//         function getNormalizedBounds(
//           page: Sketch.Page,
//           layerIndexPath: IndexPath,
//         ): Bounds {
//           const layer = Layers.access(page, layerIndexPath);
//           const transform = getLayerTransformAtIndexPath(page, layerIndexPath);
//           return createBounds(transformRect(layer.frame, transform));
//         }

//         const sortedLayerIndexPaths = layerIndexPaths.sort(
//           (a, b) =>
//             getNormalizedBounds(page, a)[sortBy] -
//             getNormalizedBounds(page, b)[sortBy],
//         );

//         let currentX = 0;
//         let currentY = 0;

//         sortedLayerIndexPaths.forEach((layerIndexPath) => {
//           const transform = getLayerTransformAtIndexPath(
//             page,
//             layerIndexPath,
//           ).invert();
//           const layer = Layers.access(
//             draft.sketch.pages[pageIndex], // access page again since we need to write to it
//             layerIndexPath,
//           );

//           switch (axis) {
//             case 'horizontal': {
//               const newOrigin = transform.applyTo({
//                 x: selectedRect.x + currentX,
//                 y: 0,
//               });
//               currentX += layer.frame.width + gapX;
//               layer.frame.x = newOrigin.x;
//               break;
//             }
//             case 'vertical': {
//               const newOrigin = transform.applyTo({
//                 x: 0,
//                 y: selectedRect.y + currentY,
//               });
//               currentY += layer.frame.height + gapY;
//               layer.frame.y = newOrigin.y;
//               break;
//             }
//           }
//         });
//       });
//     }
//     case 'alignLayers': {
//       const page = getCurrentPage(state);
//       const pageIndex = getCurrentPageIndex(state);
//       const layerIndexPaths = getSelectedLayerIndexPaths(state);
//       const selectedRect = getSelectedRect(state);
//       const [, placement] = action;

//       return produce(state, (draft) => {
//         const selectedBounds = createBounds(selectedRect);
//         const midX = selectedRect.x + selectedRect.width / 2;
//         const midY = selectedRect.y + selectedRect.height / 2;

//         layerIndexPaths.forEach((layerIndexPath) => {
//           const transform = getLayerTransformAtIndexPath(
//             page,
//             layerIndexPath,
//           ).invert();
//           const layer = Layers.access(
//             draft.sketch.pages[pageIndex], // access page again since we need to write to it
//             layerIndexPath,
//           );

//           switch (placement) {
//             case 'left': {
//               const newOrigin = transform.applyTo({
//                 x: selectedBounds.minX,
//                 y: 0,
//               });
//               layer.frame.x = newOrigin.x;
//               break;
//             }
//             case 'centerHorizontal': {
//               const newOrigin = transform.applyTo({
//                 x: midX - layer.frame.width / 2,
//                 y: 0,
//               });
//               layer.frame.x = newOrigin.x;
//               break;
//             }
//             case 'right': {
//               const newOrigin = transform.applyTo({
//                 x: selectedBounds.maxX - layer.frame.width,
//                 y: 0,
//               });
//               layer.frame.x = newOrigin.x;
//               break;
//             }
//             case 'top': {
//               const newOrigin = transform.applyTo({
//                 x: 0,
//                 y: selectedBounds.minY,
//               });
//               layer.frame.y = newOrigin.y;
//               break;
//             }
//             case 'centerVertical': {
//               const newOrigin = transform.applyTo({
//                 x: 0,
//                 y: midY - layer.frame.height / 2,
//               });
//               layer.frame.y = newOrigin.y;
//               break;
//             }
//             case 'bottom': {
//               const newOrigin = transform.applyTo({
//                 x: 0,
//                 y: selectedBounds.maxY - layer.frame.height,
//               });
//               layer.frame.y = newOrigin.y;
//               break;
//             }
//           }
//         });
//       });
//     }
//     default:
//       return state;
//   }
// }
