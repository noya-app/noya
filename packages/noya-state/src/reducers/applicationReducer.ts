import Sketch from '@sketch-hq/sketch-file-format-ts';
import { CanvasKit } from 'canvaskit';
import produce from 'immer';
import { WritableDraft } from 'immer/dist/internal';
import { uuid } from 'noya-renderer';
import { SketchFile } from 'noya-sketch-file';
import { IndexPath } from 'tree-visit';
import { KeyModifiers } from 'noya-keymap';
import * as Layers from '../layers';
import {
  findPageLayerIndexPaths,
  getCurrentComponentsTab,
  getCurrentPageIndex,
  getCurrentTab,
  getSelectedLayerIndexPaths,
  setNewPatternFill,
} from '../selectors/selectors';
import { AlignmentAction, alignmentReducer } from './alignmentReducer';
import { CanvasAction, canvasReducer } from './canvasReducer';
import { ExportAction, exportReducer } from './exportReducer';
import {
  createInitialInteractionState,
  interactionReducer,
  InteractionState,
} from './interactionReducer';
import {
  LayerPropertyAction,
  layerPropertyReducer,
} from './layerPropertyReducer';
import { LayerAction, layerReducer } from './layerReducer';
import { PageAction, pageReducer } from './pageReducer';
import { PointAction, pointReducer } from './pointReducer';
import { SetNumberMode, StyleAction, styleReducer } from './styleReducer';
import { SymbolsAction, symbolsReducer } from './symbolsReducer';
import { TextStyleAction, textStyleReducer } from './textStyleReducer';
import { ThemeAction, themeReducer } from './themeReducer';

export type { SetNumberMode };

export type WorkspaceTab = 'canvas' | 'theme';

export type ThemeTab = 'swatches' | 'textStyles' | 'layerStyles' | 'symbols';
type ThemeSelection = { ids: string[]; groupName: string };

export type controlPointType = 'curveFrom' | 'curveTo';

export type SelectedControlPoint = {
  layerId: string;
  pointIndex: number;
  controlPointType: controlPointType;
};

export type SelectedPointLists = Record<string, number[]>;

export type ApplicationState = {
  currentTab: WorkspaceTab;
  currentThemeTab: ThemeTab;
  interactionState: InteractionState;
  keyModifiers: KeyModifiers;
  selectedPage: string;
  selectedObjects: string[];
  selectedPointLists: SelectedPointLists;
  selectedControlPoint?: SelectedControlPoint;
  selectedThemeTab: Record<ThemeTab, ThemeSelection>;
  sketch: SketchFile;
};

export type Action =
  | [type: 'setTab', value: WorkspaceTab]
  | [type: 'setKeyModifier', name: keyof KeyModifiers, value: boolean]
  | PageAction
  | CanvasAction
  | LayerPropertyAction
  | LayerAction
  | AlignmentAction
  | StyleAction
  | TextStyleAction
  | ThemeAction
  | SymbolsAction
  | ExportAction
  | PointAction;

export function applicationReducer(
  state: ApplicationState,
  action: Action,
  CanvasKit: CanvasKit,
): ApplicationState {
  switch (action[0]) {
    case 'setKeyModifier':
      const [, name, value] = action;
      return produce(state, (draft) => {
        draft.keyModifiers[name] = value;
      });
    case 'setTab': {
      const [, value] = action;
      return produce(state, (draft) => {
        draft.currentTab = value;
        draft.interactionState = interactionReducer(draft.interactionState, [
          'reset',
        ]);
      });
    }
    case 'selectPage':
    case 'addPage':
    case 'renamePage':
    case 'duplicatePage':
    case 'deletePage':
    case 'movePage': {
      return pageReducer(state, action);
    }
    case 'insertArtboard':
    case 'addDrawnLayer':
    case 'addShapePathLayer':
    case 'addSymbolLayer':
    case 'addPointToPath':
    case 'insertBitmap':
    case 'pan':
    case 'insertBitmap':
    case 'pan':
    case 'interaction':
      return canvasReducer(state, action, CanvasKit);
    case 'setLayerVisible':
    case 'setLayerIsLocked':
    case 'setExpandedInLayerList':
    case 'setFixedRadius':
    case 'setLayerX':
    case 'setLayerY':
    case 'setLayerWidth':
    case 'setLayerHeight':
    case 'setLayerRotation':
    case 'setIsClosed':
    case 'setIsFlippedHorizontal':
    case 'setIsFlippedVertical':
    case 'setHasClippingMask':
    case 'setShouldBreakMaskChain':
      return layerPropertyReducer(state, action);
    case 'groupLayer':
    case 'deleteLayer':
    case 'moveLayer':
    case 'selectLayer':
    case 'selectAllLayers':
    case 'ungroupLayer':
    case 'createSymbol':
    case 'detachSymbol':
    case 'deleteSymbol':
    case 'duplicateLayer':
      return layerReducer(state, action);
    case 'distributeLayers':
    case 'alignLayers': {
      return alignmentReducer(state, action);
    }
    case 'addNewBorder':
    case 'addNewFill':
    case 'addNewShadow':
    case 'setBorderEnabled':
    case 'setFillEnabled':
    case 'setShadowEnabled':
    case 'deleteBorder':
    case 'deleteFill':
    case 'deleteShadow':
    case 'moveBorder':
    case 'moveFill':
    case 'moveShadow':
    case 'deleteDisabledBorders':
    case 'deleteDisabledFills':
    case 'deleteDisabledShadows':
    case 'setBorderColor':
    case 'setFillColor':
    case 'setShadowColor':
    case 'setBorderWidth':
    case 'setFillOpacity':
    case 'setOpacity':
    case 'setShadowX':
    case 'setShadowY':
    case 'setShadowBlur':
    case 'setBorderPosition':
    case 'setShadowSpread':
    case 'setFillFillType':
    case 'setBorderFillType':
    case 'setFillGradientColor':
    case 'setFillGradientPosition':
    case 'setFillGradientType':
    case 'setFillGradient':
    case 'addFillGradientStop':
    case 'deleteFillGradientStop':
    case 'setBorderGradient':
    case 'setBorderGradientColor':
    case 'setBorderGradientPosition':
    case 'setBorderGradientType':
    case 'addBorderGradientStop':
    case 'deleteBorderGradientStop':
    case 'setColorControlsEnabled':
    case 'setHue':
    case 'setSaturation':
    case 'setBrightness':
    case 'setContrast':
    case 'setPatternFillType':
    case 'setPatternTileScale':
    case 'setFillImage':
    case 'setFillContextSettingsOpacity': {
      if (getCurrentTab(state) === 'canvas') {
        const pageIndex = getCurrentPageIndex(state);
        const layerIndexPaths = getSelectedLayerIndexPaths(state);

        return produce(state, (draft) => {
          accessPageLayers(draft, pageIndex, layerIndexPaths).forEach(
            (layer) => {
              if (!layer.style) return;
              if (
                action[0] === 'setFillFillType' &&
                action[2] === Sketch.FillType.Pattern &&
                layer.style.fills
              )
                setNewPatternFill(layer.style.fills, action[1], draft);

              layer.style = styleReducer(layer.style, action);
            },
          );
        });
      } else {
        const ids = state.selectedThemeTab.layerStyles.ids;

        const layerIndexPathsWithSharedStyle = findPageLayerIndexPaths(
          state,
          (layer) =>
            layer.sharedStyleID !== undefined &&
            ids.includes(layer.sharedStyleID),
        );

        const currentComponentsTab = getCurrentComponentsTab(state);

        const selectedIds = state.selectedThemeTab[currentComponentsTab].ids;

        return produce(state, (draft) => {
          const styles =
            currentComponentsTab === 'layerStyles'
              ? draft.sketch.document.layerStyles?.objects
              : draft.sketch.document.layerTextStyles?.objects ?? [];

          styles.forEach((style) => {
            if (!selectedIds.includes(style.do_objectID)) return;

            style.value = styleReducer(style.value, action);

            if (
              action[0] === 'setFillFillType' &&
              action[2] === Sketch.FillType.Pattern &&
              style.value.fills
            )
              setNewPatternFill(style.value.fills, action[1], draft);

            layerIndexPathsWithSharedStyle.forEach((layerPath) =>
              accessPageLayers(
                draft,
                layerPath.pageIndex,
                layerPath.indexPaths,
              ).forEach((layer) => {
                layer.style = produce(style.value, (style) => {
                  style.do_objectID = uuid();
                  return style;
                });
              }),
            );
          });
        });
      }
    }
    case 'setTextColor':
    case 'setTextFontName':
    case 'setTextFontSize':
    case 'setTextLineSpacing':
    case 'setTextLetterSpacing':
    case 'setTextParagraphSpacing':
    case 'setTextHorizontalAlignment':
    case 'setTextVerticalAlignment':
    case 'setTextAlignment':
    case 'setTextDecoration':
    case 'setTextTransform':
      return textStyleReducer(state, action);
    case 'setAdjustContentOnResize':
    case 'setHasBackgroundColor':
    case 'setBackgroundColor':
    case 'setIncludeBackgroundColorInExport':
    case 'setIncludeBackgroundColorInInstance':
    case 'setLayoutAxis':
    case 'setLayoutAnchor':
    case 'setMinWidth':
    case 'setAllowsOverrides':
    case 'onSetOverrideProperty':
    case 'setInstanceSymbolSource':
    case 'goToSymbolSource':
    case 'setOverrideValue':
      return symbolsReducer(state, action);
    case 'setExportScale':
    case 'setExportName':
    case 'setExportFileFormat':
    case 'setExportNamingScheme':
    case 'addExportFormat':
    case 'deleteExportFormat':
      const pageIndex = getCurrentPageIndex(state);
      const layerIndexPaths = getSelectedLayerIndexPaths(state);

      return produce(state, (draft) => {
        accessPageLayers(draft, pageIndex, layerIndexPaths).forEach((layer) => {
          layer.exportOptions = exportReducer(
            layer.exportOptions,
            action,
            layer.frame,
          );
        });
      });
    case 'setPointCurveMode':
    case 'setPointCornerRadius':
    case 'setPointX':
    case 'setPointY':
    case 'setControlPointX':
    case 'setControlPointY':
    case 'selectPoint':
    case 'selectControlPoint':
      return pointReducer(state, action, CanvasKit);
    default:
      return themeReducer(state, action);
  }
}

/*
 * Get an array of all layers using as few lookups as possible on the state tree.
 *
 * Immer will duplicate any objects we access within a produce method, so we
 * don't want to walk every layer, since that would duplicate all of them.
 */
export function accessPageLayers(
  state: WritableDraft<ApplicationState>,
  pageIndex: number,
  layerIndexPaths: IndexPath[],
): Sketch.AnyLayer[] {
  return layerIndexPaths.map((layerIndex) => {
    return Layers.access(state.sketch.pages[pageIndex], layerIndex);
  });
}

export function createInitialState(sketch: SketchFile): ApplicationState {
  if (sketch.pages.length === 0) {
    throw new Error('Invalid Sketch file - no pages');
  }

  return {
    currentTab: 'canvas',
    currentThemeTab: 'swatches',
    interactionState: createInitialInteractionState(),
    keyModifiers: {
      altKey: false,
      ctrlKey: false,
      metaKey: false,
      shiftKey: false,
    },
    selectedPage: sketch.pages[0].do_objectID,
    selectedObjects: [],
    selectedPointLists: {},
    selectedControlPoint: undefined,
    selectedThemeTab: {
      swatches: { ids: [], groupName: '' },
      layerStyles: { ids: [], groupName: '' },
      textStyles: { ids: [], groupName: '' },
      symbols: { ids: [], groupName: '' },
    },
    sketch,
  };
}
