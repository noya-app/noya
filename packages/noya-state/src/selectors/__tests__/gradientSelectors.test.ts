import Sketch from 'noya-file-format';
import { SketchModel } from 'noya-sketch-model';
import { createInitialState } from '../../reducers/applicationReducer';
import { createSketchFile } from '../../sketchFile';
import {
  getAngularGradientCircle,
  getSelectedGradientStopPoints,
} from '../gradientSelectors';

const gradientStops = [
  SketchModel.gradientStop({
    position: 0,
    color: SketchModel.color({ red: 1, green: 1, blue: 1, alpha: 1 }),
  }),
  SketchModel.gradientStop({
    position: 1,
    color: SketchModel.color({ red: 0, green: 0, blue: 0, alpha: 1 }),
  }),
  SketchModel.gradientStop({
    position: 0.5,
    color: SketchModel.color({
      red: 0.5,
      green: 0.5,
      blue: 0.5,
      alpha: 0.5,
    }),
  }),
];

const angularGradientStyle = SketchModel.style({
  fills: [
    SketchModel.fill({
      fillType: Sketch.FillType.Gradient,
      gradient: SketchModel.gradient({
        gradientType: Sketch.GradientType.Angular,
        stops: gradientStops,
      }),
    }),
  ],
});

const linearGradientStyle = SketchModel.style({
  fills: [
    SketchModel.fill({
      fillType: Sketch.FillType.Gradient,
      gradient: SketchModel.gradient({
        stops: gradientStops,
      }),
    }),
  ],
});

describe('set angular gradient circle', () => {
  const rectangle = SketchModel.rectangle({
    frame: SketchModel.rect({
      x: 0,
      y: 0,
      width: 100,
      height: 100,
    }),
    style: angularGradientStyle,
  });

  test('top layer', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle] })),
    );
    state.selectedGradient = {
      layerId: rectangle.do_objectID,
      fillIndex: 0,
      stopIndex: 0,
      styleType: 'fills',
    };

    expect(getAngularGradientCircle(state)).toEqual({
      center: { x: 50, y: 50 },
      radius: 50,
      rotation: 0,
    });
  });

  test('flipped layer horizontal', () => {
    const flippedRectangle = { ...rectangle, isFlippedHorizontal: true };

    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [flippedRectangle] })),
    );
    state.selectedGradient = {
      layerId: flippedRectangle.do_objectID,
      fillIndex: 0,
      stopIndex: 0,
      styleType: 'fills',
    };

    expect(getAngularGradientCircle(state)).toEqual({
      center: { x: 50, y: 50 },
      radius: 50,
      rotation: Math.PI,
    });
  });

  test('flipped layer vertical', () => {
    const flippedRectangle = { ...rectangle, isFlippedVertical: true };

    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [flippedRectangle] })),
    );
    state.selectedGradient = {
      layerId: flippedRectangle.do_objectID,
      fillIndex: 0,
      stopIndex: 0,
      styleType: 'fills',
    };

    expect(getAngularGradientCircle(state)).toEqual({
      center: { x: 50, y: 50 },
      radius: 50,
      rotation: 0,
    });
  });

  test('rotated layer', () => {
    const rotatedRectangle = { ...rectangle, rotation: 1 };

    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rotatedRectangle] })),
    );
    state.selectedGradient = {
      layerId: rectangle.do_objectID,
      fillIndex: 0,
      stopIndex: 0,
      styleType: 'fills',
    };

    expect(getAngularGradientCircle(state)).toMatchSnapshot();
  });

  test('no selected gradient', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle] })),
    );
    state.selectedGradient = undefined;

    expect(getAngularGradientCircle(state)).toEqual(undefined);
  });

  test('layer in artboard', () => {
    const artboard = SketchModel.artboard({
      frame: SketchModel.rect({
        x: 450,
        y: 450,
        width: 300,
        height: 300,
      }),
      layers: [rectangle],
    });

    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [artboard] })),
    );
    state.selectedGradient = {
      layerId: rectangle.do_objectID,
      fillIndex: 0,
      stopIndex: 0,
      styleType: 'fills',
    };

    expect(getAngularGradientCircle(state)).toEqual({
      center: { x: 500, y: 500 },
      radius: 50,
      rotation: 0,
    });
  });

  test('layer in group in artboard', () => {
    const artboard = SketchModel.artboard({
      frame: SketchModel.rect({
        x: 450,
        y: 450,
        width: 300,
        height: 300,
      }),
      layers: [
        SketchModel.group({
          do_objectID: 'g1',
          frame: SketchModel.rect({ x: 50, y: 50, width: 100, height: 100 }),
          layers: [rectangle],
        }),
      ],
    });

    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [artboard] })),
    );

    state.selectedGradient = {
      layerId: rectangle.do_objectID,
      fillIndex: 0,
      stopIndex: 0,
      styleType: 'fills',
    };

    expect(getAngularGradientCircle(state)).toEqual({
      center: { x: 550, y: 550 },
      radius: 50,
      rotation: 0,
    });
  });
});

describe('get Gradient Point', () => {
  const linearGradientRectangle = SketchModel.rectangle({
    frame: SketchModel.rect({
      x: 0,
      y: 0,
      width: 100,
      height: 100,
    }),
    style: linearGradientStyle,
  });

  test('linear gradient / radial gradient', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [linearGradientRectangle] })),
    );

    state.selectedGradient = {
      layerId: linearGradientRectangle.do_objectID,
      fillIndex: 0,
      stopIndex: 0,
      styleType: 'fills',
    };

    expect(getSelectedGradientStopPoints(state)).toMatchSnapshot();
  });

  test('angular gradient', () => {
    const rectangle = {
      ...linearGradientRectangle,
      style: angularGradientStyle,
    };

    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle] })),
    );

    state.selectedGradient = {
      layerId: rectangle.do_objectID,
      fillIndex: 0,
      stopIndex: 0,
      styleType: 'fills',
    };

    expect(getSelectedGradientStopPoints(state)).toMatchSnapshot();
  });

  test('sorted points', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [linearGradientRectangle] })),
    );

    state.selectedGradient = {
      layerId: linearGradientRectangle.do_objectID,
      fillIndex: 0,
      stopIndex: 0,
      styleType: 'fills',
    };

    expect(getSelectedGradientStopPoints(state, true)).toMatchSnapshot();
  });

  test('no selected gradient', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [linearGradientRectangle] })),
    );

    state.selectedGradient = undefined;

    expect(getSelectedGradientStopPoints(state)).toEqual(undefined);
  });

  test('layer in artboard', () => {
    const artboard = SketchModel.artboard({
      frame: SketchModel.rect({
        x: 450,
        y: 450,
        width: 300,
        height: 300,
      }),
      layers: [linearGradientRectangle],
    });

    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [artboard] })),
    );
    state.selectedGradient = {
      layerId: linearGradientRectangle.do_objectID,
      fillIndex: 0,
      stopIndex: 0,
      styleType: 'fills',
    };

    expect(getSelectedGradientStopPoints(state, true)).toMatchSnapshot();
  });

  test('layer in group in artboard', () => {
    const artboard = SketchModel.artboard({
      frame: SketchModel.rect({
        x: 450,
        y: 450,
        width: 300,
        height: 300,
      }),
      layers: [
        SketchModel.group({
          do_objectID: 'g1',
          frame: SketchModel.rect({ x: 50, y: 50, width: 100, height: 100 }),
          layers: [linearGradientRectangle],
        }),
      ],
    });

    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [artboard] })),
    );

    state.selectedGradient = {
      layerId: linearGradientRectangle.do_objectID,
      fillIndex: 0,
      stopIndex: 0,
      styleType: 'fills',
    };

    expect(getSelectedGradientStopPoints(state, true)).toMatchSnapshot();
  });
});
