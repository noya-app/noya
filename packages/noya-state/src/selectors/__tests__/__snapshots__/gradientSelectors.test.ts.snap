// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get Gradient Point angular gradient 1`] = `
Array [
  Object {
    "color": Object {
      "_class": "color",
      "alpha": 1,
      "blue": 1,
      "green": 1,
      "red": 1,
    },
    "point": Object {
      "x": 100,
      "y": 50,
    },
  },
  Object {
    "color": Object {
      "_class": "color",
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 0,
    },
    "point": Object {
      "x": 100,
      "y": 49.999999999999986,
    },
  },
  Object {
    "color": Object {
      "_class": "color",
      "alpha": 0.5,
      "blue": 0.5,
      "green": 0.5,
      "red": 0.5,
    },
    "point": Object {
      "x": 0,
      "y": 50.00000000000001,
    },
  },
]
`;

exports[`get Gradient Point layer in artboard 1`] = `
Array [
  Object {
    "color": Object {
      "_class": "color",
      "alpha": 1,
      "blue": 1,
      "green": 1,
      "red": 1,
    },
    "point": Object {
      "x": 500,
      "y": 450,
    },
  },
  Object {
    "color": Object {
      "_class": "color",
      "alpha": 0.5,
      "blue": 0.5,
      "green": 0.5,
      "red": 0.5,
    },
    "point": Object {
      "x": 500,
      "y": 500,
    },
  },
  Object {
    "color": Object {
      "_class": "color",
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 0,
    },
    "point": Object {
      "x": 500,
      "y": 550,
    },
  },
]
`;

exports[`get Gradient Point layer in group in artboard 1`] = `
Array [
  Object {
    "color": Object {
      "_class": "color",
      "alpha": 1,
      "blue": 1,
      "green": 1,
      "red": 1,
    },
    "point": Object {
      "x": 550,
      "y": 500,
    },
  },
  Object {
    "color": Object {
      "_class": "color",
      "alpha": 0.5,
      "blue": 0.5,
      "green": 0.5,
      "red": 0.5,
    },
    "point": Object {
      "x": 550,
      "y": 550,
    },
  },
  Object {
    "color": Object {
      "_class": "color",
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 0,
    },
    "point": Object {
      "x": 550,
      "y": 600,
    },
  },
]
`;

exports[`get Gradient Point linear gradient / radial gradient 1`] = `
Array [
  Object {
    "color": Object {
      "_class": "color",
      "alpha": 1,
      "blue": 1,
      "green": 1,
      "red": 1,
    },
    "point": Object {
      "x": 50,
      "y": 0,
    },
  },
  Object {
    "color": Object {
      "_class": "color",
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 0,
    },
    "point": Object {
      "x": 50,
      "y": 100,
    },
  },
  Object {
    "color": Object {
      "_class": "color",
      "alpha": 0.5,
      "blue": 0.5,
      "green": 0.5,
      "red": 0.5,
    },
    "point": Object {
      "x": 50,
      "y": 50,
    },
  },
]
`;

exports[`get Gradient Point sorted points 1`] = `
Array [
  Object {
    "color": Object {
      "_class": "color",
      "alpha": 1,
      "blue": 1,
      "green": 1,
      "red": 1,
    },
    "point": Object {
      "x": 50,
      "y": 0,
    },
  },
  Object {
    "color": Object {
      "_class": "color",
      "alpha": 0.5,
      "blue": 0.5,
      "green": 0.5,
      "red": 0.5,
    },
    "point": Object {
      "x": 50,
      "y": 50,
    },
  },
  Object {
    "color": Object {
      "_class": "color",
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 0,
    },
    "point": Object {
      "x": 50,
      "y": 100,
    },
  },
]
`;

exports[`set angular gradient circle rotated layer 1`] = `
Object {
  "center": Object {
    "x": 50,
    "y": 50,
  },
  "radius": 50,
  "rotation": -0.017453292519943295,
}
`;
